### PR TITLE
nhrp: T4905: Rewritten nhrp op-mode in new style

### PR DESCRIPTION
**Pull request by aapostoliuk**
## Change Summary

1. Formatted output of 'show nhrp' commands to table view
2. Rewritten nhrp op-mode in new style

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://phabricator.vyos.net/T4905

## Component(s) name

nhrp, op-mode

## Proposed changes

1. Formatted output of 'show nhrp' commands to table view
```
vyos@vyos:~$ show nhrp interface
Status: ok
Interface    Index    Protocol-Address    Flags                         Holding-Time    Route-Table    GRE-Key    MTU    NBMA-MTU    NBMA-Address
-----------  -------  ------------------  ----------------------------  --------------  -------------  ---------  -----  ----------  --------------
tun100       9        10.100.100.2/24     shortcut redirect configured  300             254            1          1476   0           192.168.192.22
erspan0      8
gretap0      7
gre0         6
eth3         5
eth2         4
eth1         3
eth0         2
lo           1
vyos@vyos:~$ show nhrp tunnel
Status: ok
Interface    Type    Protocol-Address    Alias-Address    Flags    NBMA-Address
-----------  ------  ------------------  ---------------  -------  --------------
tun100       local   10.100.100.255/32   10.100.100.2     up
tun100       local   10.100.100.2/32                      up
tun100       static  10.100.100.1/24                      up       192.168.192.21
```

2. Rewritten nhrp op-mode in new style

## Checklist:

- [x] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [x] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [x] My commit headlines contain a valid Task id
- [ ] My change requires a change to the documentation
- [ ] I have updated the documentation accordingly

### CODE REVIEW

File: src/op_mode/nhrp.py [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+
+def show_interface(raw: bool):
+    """
+    Command 'show nhrp interface'
+    :param raw: if API
+    :type raw: bool
+    """
+    if not process_named_running('opennhrp'):
+        raise vyos.opmode.UnsupportedOperation('OpenNHRP is not running.')
```
Comment (dmbaturin) [inline src/op_mode/nhrp.py]: We have a dedicated exception specially for this situation: [`UnconfiguredSubsystem`](https://github.com/vyos/vyos-1x/blob/current/python/vyos/opmode.py#L32-L36). `UnsupportedOperation` is for operations that the current VyOS version doesn't support at all, e.g., when a tool lacks machine-readable output for some of the commands.

File: src/op_mode/nhrp.py [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+
+
+def show_tunnel(raw: bool):
+    """
+    Command 'show nhrp tunnel'
+    :param raw: if API
+    :type raw: bool
+    """
+    if not process_named_running('opennhrp'):
+        raise vyos.opmode.UnsupportedOperation('OpenNHRP is not running.')
```
Comment (dmbaturin) [inline src/op_mode/nhrp.py]: Should also be `UnconfiguredSubsystem`.

File: src/op_mode/nhrp.py [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+    """
+    Format string returned from CMD to API list
+    :param output_string: String received by CMD
+    :type output_string: str
+    :return: list for API
+    :rtype: list
+    """
+    formatted_list: list = []
+    output_list: list = output_string.split('\n\n')
+    for list_a in output_list:
```
Comment (dmbaturin) [inline src/op_mode/nhrp.py:53]: We have a function for converting colon-separated lists to dicts: [vyos.util.colon_separated_to_dict](https://github.com/vyos/vyos-1x/blob/current/python/vyos/util.py#L320).

File: src/op_mode/nhrp.py [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+    """
+    status: dict = output_list.pop(0)
+    if 'Status' in status:
+        print(f"Status: {status['Status']}")
+    output: str = tabulate.tabulate(output_list, headers='keys',
+                                    numalign="left")
+    return output
+
+
+def get_formatted_list(output_string: str) -> list:
```
Comment (dmbaturin) [inline src/op_mode/nhrp.py]: First, I believe for such relatively obscure formats we should always include output samples in the code, or at least describe what the format is.

Second, the current JSON output is very annoying to work with for any client:

```
[{'Status': 'ok'},
 {'Interface': 'tun0', 'Type': 'local', 'Protocol-Address': '10.4.0.255/32', 'Alias-Address': '10.4.0.1', 'Flags': 'up'}, {'Interface': 'tun0', 'Type': 'local', 'Protocol-Address': '10.4.0.1/32', 'Flags': 'up'}, {'Interface': 'tun0', 'Type': 'local', 'Protocol-Address': '10.4.0.255/32', 'Alias-Address': '10.4.0.1', 'Flags': 'up'}, {'Interface': 'tun0', 'Type': 'local', 'Protocol-Address': '10.4.0.1/32', 'Flags': 'up'}]
```

The list contains objects of different types: the status object and a bunch of route objects, so any client code has to separate them, it cannot just loop through the routes.

At the very least we should do something like this:

```
{ 'status': 'Ok",
  'routes': [{'Interface': 'tun0', ...}, ...]
```

File: src/op_mode/nhrp.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+
+import sys
+import tabulate
+import vyos.opmode
+
+from vyos.util import cmd
+from vyos.util import process_named_running
+
+
+def get_formatted_output(output_list: list) -> str:
```
Comment (dmbaturin) [inline src/op_mode/nhrp.py]: These function names should probably be underscored (`_get_formatted_output`) to mark them non-public.

**Review by dmbaturin — CHANGES_REQUESTED** [review]

File: src/op_mode/nhrp.py (hunk 1 above)
Comment (aapostoliuk) [reply src/op_mode/nhrp.py]: done

File: src/op_mode/nhrp.py (hunk 2 above)
Comment (aapostoliuk) [reply src/op_mode/nhrp.py]: done

File: src/op_mode/nhrp.py (hunk 3 above)
Comment (aapostoliuk) [reply src/op_mode/nhrp.py:53]: done

File: src/op_mode/nhrp.py (hunk 5 above)
Comment (aapostoliuk) [reply src/op_mode/nhrp.py]: done

File: src/op_mode/nhrp.py (hunk 4 above)
Comment (aapostoliuk) [reply src/op_mode/nhrp.py]: done

**Comment by sever-sever** [comment]
What about expires time?
```
Interface: tun0
Type: dynamic
Protocol-Address: 10.214.3.23/32
NBMA-Address: 10.10.3.23
Flags: up
Expires-In: 97:04
```

**Comment by aapostoliuk** [comment]
> What about expires time?
> 
> ```
> Interface: tun0
> Type: dynamic
> Protocol-Address: 10.214.3.23/32
> NBMA-Address: 10.10.3.23
> Flags: up
> Expires-In: 97:04
> ```

It will appear if this key is in the dictionary.

**Review by dmbaturin — APPROVED** [review]

**Pull request merged**